File: godfather/roles/mixins/notarget.py
from godfather.roles.base import Role
from godfather.errors import PhaseChangeError


class NoTarget(Role):
    action = ''
    action_text = ''
    can_block = True
    can_transport = True

    async def on_night(self, bot, player, game):
        output = f'It is now night {game.cycle}. Use the {bot.command_prefix}{self.action} command to {self.action_text}. ' \
            + f'Use {bot.command_prefix}noaction to stay home.\n'
        await player.user.send(output)

    async def set_up(self, actions, player, target):
        pass

    async def run_action(self, actions, player, target):
        pass

    async def tear_down(self, actions, player, target):
        pass

    async def on_pm_command(self, ctx, game, player, args):
        command = args[0]
        can_do, reason = self.can_do_action(game)
        if not can_do:
            return await ctx.send(f'You cannot use your action today. {reason}')

        if command == 'noaction':
            for action in game.night_actions:
                if action['player'].user.id == player.user.id:
                    game.night_actions.remove(action)

            game.night_actions.add_action({
                'action': None,
                'player': player,
                'priority': 0
            })
            return await ctx.send('You decided to stay home tonight.')

        game.night_actions.add_action({
            'action': self.action,
            'player': player,
            'priority': self.action_priority,
            'target': player,
            'can_block': self.can_block,
            'can_transport': self.can_transport
        })
        await ctx.send('You have decided to {} tonight.'.format(self.action))

        if len(game.players.filter(action_only=True)) == len(game.night_actions):
            try:
                await game.increment_phase()
            except Exception as exc:
                raise PhaseChangeError(None, *exc.args)

File: src/lib/mafia/mixins/NoTarget.ts
import { DEFAULT_ACTION_FLAGS } from '@lib/constants';
// @ts-expect-error (1371) Bypass not being able to type import default and named
import NightActionsManager, { NightActionPriority } from '@mafia/managers/NightActionsManager';
import { Phase } from '@mafia/structures/Game';
import Role from '@mafia/structures/Role';
import { PREFIX } from '@root/config';
import type { Awaited } from '@sapphire/framework';
import { removeArrayItem } from '@util/utils';
import type { Message } from 'discord.js';

class NoTarget extends Role {
	public actionPhase = Phase.Night;

	public async onNight() {
		const { game } = this.player;
		const actionText = [
			game.t('roles/global:nightActionPromptNoTarget', {
				prefix: PREFIX,
				cycle: game.cycle,
				action: this.action,
				actionText: this.actionText
			}),
			this.extraNightContext
		]
			.filter((text) => text !== null)
			.join('\n');
		await this.player.user.send(actionText);
	}

	public async onPmCommand(message: Message, command: string) {
		// day commands use a completely different action flow
		const { check, reason } = this.canUseAction();
		if (!check) throw this.game.t('roles/global:actionBlocked', { reason });

		if (!this.possibleActions.includes(command)) return;

		removeArrayItem(this.game.nightActions, (action) => action.actor === this.player);

		switch (command) {
			case 'cancel':
				return this.player.user.send(this.game.t('roles/global:actionCancelled'));
			case 'noaction': {
				await this.game.nightActions.addAction({
					action: undefined,
					actor: this.player,
					priority: this.priority
				});
				return this.player.user.send(this.game.t('roles/global:noAction'));
			}
			default:
				await this.player.game.nightActions.addAction({
					action: this.action,
					actor: this.player,
					priority: this.priority,
					flags: this.flags ?? DEFAULT_ACTION_FLAGS
				});
		}

		return this.player.user.send(this.game.t('roles/global:noTargetConfirmation', { gerund: this.actionGerund }));
	}

	private get possibleActions() {
		return [this.action, 'noaction'];
	}

	public canUseAction() {
		return { check: true, reason: '' };
	}

	// eslint-disable-next-line @typescript-eslint/no-unused-vars
	public setUp(actions: NightActionsManager): Awaited<any> {
		// noop
	}

	// eslint-disable-next-line @typescript-eslint/no-unused-vars
	public runAction(actions: NightActionsManager): Awaited<any> {
		// noop
	}

	// eslint-disable-next-line @typescript-eslint/no-unused-vars
	public tearDown(actions: NightActionsManager): Awaited<any> {
		// noop
	}

	// eslint-disable-next-line @typescript-eslint/class-literal-property-style
	public get extraNightContext(): string | null {
		return null;
	}
}

interface NoTarget {
	action: string;
	actionGerund: string;
	actionText: string;
	flags?: {
		canBlock: boolean;
		canTransport: boolean;
		canVisit: boolean;
	};
	priority: NightActionPriority;
}

export default NoTarget;
